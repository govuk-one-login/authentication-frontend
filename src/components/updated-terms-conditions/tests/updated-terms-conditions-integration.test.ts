import { beforeEach, describe } from "mocha";
import { sinon, request } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import * as cheerio from "cheerio";
import decache from "decache";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants.js";
import type { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
import { buildMfaMethods } from "../../../../test/helpers/mfa-helper.js";

describe("Integration:: updated-terms-code", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    const sessionMiddleware = await import(
      "../../../middleware/session-middleware.js"
    );

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
        res.locals.clientSessionId = "tDy103saszhcxbQq0-mjdzU33d";
        res.locals.persistentSessionId = "dips-123456-abc";

        req.session.user = {
          email: "test@test.com",
          mfaMethods: buildMfaMethods({ phoneNumber: "7867" }),
          journey: getPermittedJourneyForPath(
            PATH_NAMES.UPDATED_TERMS_AND_CONDITIONS
          ),
        };

        next();
      });

    app = await (await import("../../../app.js")).createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL;

    await request(
      app,
      (test) => test.get(PATH_NAMES.UPDATED_TERMS_AND_CONDITIONS),
      { expectAnalyticsPropertiesMatchSnapshot: false }
    ).then((res) => {
      const $ = cheerio.load(res.text);
      token = $("[name=_csrf]").val();
      cookies = res.headers["set-cookie"];
    });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should return update terms and conditions page", async () => {
    await request(app, (test) =>
      test
        .get(PATH_NAMES.UPDATED_TERMS_AND_CONDITIONS)
        .expect(HTTP_STATUS_CODES.OK)
    );
  });

  it("should return error when csrf not present", async () => {
    await request(app, (test) =>
      test
        .post(PATH_NAMES.UPDATED_TERMS_AND_CONDITIONS)
        .type("form")
        .send({
          termsAndConditionsResult: "reject",
        })
        .expect(403)
    );
  });

  it("should redirect to /auth_code when terms accepted", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.UPDATE_PROFILE)
      .once()
      .reply(HTTP_STATUS_CODES.NO_CONTENT);

    await request(app, (test) =>
      test
        .post(PATH_NAMES.UPDATED_TERMS_AND_CONDITIONS)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          termsAndConditionsResult: "accept",
        })
        .expect("Location", PATH_NAMES.AUTH_CODE)
        .expect(302)
    );
  });
});
