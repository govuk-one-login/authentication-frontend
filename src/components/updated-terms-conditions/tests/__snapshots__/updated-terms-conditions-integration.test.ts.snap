// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration:: updated-terms-code should return error when csrf not present 1`] = `
Object {
  "contentId": undefined,
  "taxonomyLevel1": undefined,
  "taxonomyLevel2": undefined,
  "taxonomyLevel3": undefined,
  "taxonomyLevel4": undefined,
  "taxonomyLevel5": undefined,
}
`;

exports[`Integration:: updated-terms-code should return update terms and conditions page 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "authentication",
  "taxonomyLevel2": "",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
